### ISSUE TITLE
Custom optimizers

### ISSUE THREAD

**Issue by zkourouma** [thumbor/thumbor#393]
the jpegtran optimizer that you guys implemented works great!

i've been having some issues with implementing a png one with [optipng](http://fossies.org/linux/privat/optipng-0.7.5.tar.gz/) (below). Its very simple and i can see the verbose output so i know its taking the input/output files the way i want (naturally the bin works locally with pngs) so i'm not sure if i'm missing something or if anyone else had an issue (tried pngcrush and pngquant as well).

i pdb'd through after the command was called and saw that the file is being written so that i can see both files in Finder. however, when i call `output_file.seek(0)` and then `output_file.read()` it returns an empty string. 

i'm using chrome v 39.0.2171.71 (64-bit). here's the network tab from dev tools
![screen shot 2014-12-02 at 11 05 52](https://cloud.githubusercontent.com/assets/3931509/5265805/379e95f8-7a13-11e4-8926-b92beac2c1d6.png)

```python
import os

from thumbor.optimizers import BaseOptimizer

class Optimizer(BaseOptimizer):

    def should_run(self, image_extension, buffer):
        return 'png' in image_extension

    def optimize(self, buffer, input_file, output_file):
        png_path = self.context.config.OPTIPNG_PATH
        command = '%s -keep -clobber -preserve -out %s %s ' % (
            png_path,
            output_file,
            input_file,
        )
        os.system(command)
```

i get the same issue with the pngquant implementation below

```python
import os

from thumbor.optimizers import BaseOptimizer

class Optimizer(BaseOptimizer):

    def should_run(self, image_extension, buffer):
        return 'png' in image_extension

    def optimize(self, buffer, input_file, output_file):
        png_path = self.context.config.PNGQUANT_PATH
        command = '%s --force --verbose --output %s %s ' % (
            png_path,
            output_file,
            input_file,
        )
        os.system(command)
```

**Comment by chentschel** [comment]
Hi guys.!. 

I've implemented a pngquanta optimizer and had this issue also. 

Here's the fix i have to make. 

```
class BaseOptimizer(object):
    def __init__(self, context):
        self.context = context

    def should_run(self, image_extension, buffer):
        return True

    def run_optimizer(self, image_extension, buffer):
        if not self.should_run(image_extension, buffer):
            return buffer

        ifile = NamedTemporaryFile(delete=False)
        ofile = NamedTemporaryFile(delete=False)
        try:
            ifile.write(buffer)
            ifile.close()
            ofile.close()

            self.optimize(buffer, ifile.name, ofile.name)

            ofile = open(ofile.name, 'rb') #reopen with file thats been changed with the optimizer
            return ofile.read()

        finally:
            os.unlink(ifile.name)
            os.unlink(ofile.name)
```

and the optimizer. 

```
import os

from thumbor.optimizers import BaseOptimizer

class Optimizer(BaseOptimizer):

    def should_run(self, image_extension, buffer):
        return 'png' in image_extension

    def optimize(self, buffer, input_file, output_file):
        pngquant_path = self.context.config.PNGQUANT_PATH
        command = '%s -f -o %s %s' % (
            pngquant_path,
            output_file,
            input_file
        )
        os.system(command)
```

**Comment by heynemann** [comment]
@guilhermef can you apply this patch to thumbor to improve our user of optimizers?

**Comment by guilhermef** [comment]
@chentschel I've just applied your patch.
After you're done testing I'll release a new version, ok ?

**Comment by chentschel** [comment]
Great @guilhermef !. I've tested it, and its in use in our prod. servers. http://oony.com.

I've found an issue with png optimizer implementation (pngquant) as i've shared here. --quality parameter It does not output any file if parameters are not carefully implemented, so sometimes could return a 0 byte len file resulting in thumbor optimizer class returning a 0 len file.

CH

**Comment by guilhermef** [comment]
Guys, I've just released the 4.9.0.
Sorry for the delay.
Thanks.
